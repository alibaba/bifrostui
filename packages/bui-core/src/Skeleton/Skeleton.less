.bui-skeleton {
  --border-radius: var(--bui-shape-radius-label);
  display: block;

  &-animation-wave {
    background: linear-gradient(
      90deg,
      rgba(190, 190, 190, 0.2) 25%,
      rgba(129, 129, 129, 0.24) 37%,
      rgba(190, 190, 190, 0.2) 63%
    );
    background-size: 400% 100%;
    animation: bui-skeleton-wave 1.4s ease infinite;
  }

  &-animation-pulse {
    background: rgba(190, 190, 190, 0.35);
    animation: bui-skeleton-pulse 1.4s ease infinite;
  }

  &-text {
    height: auto;
    transform-origin: 0 55%;
    transform: scale(1, 0.6);
    border-radius: var(--border-radius);

    &::before {
      content: '\00a0';
    }
  }

  &-rectangular {
    height: 1.2em;
  }

  &-rounded {
    height: 1.2em;
    border-radius: var(--border-radius);
  }

  &-circular {
    border-radius: 50%;
  }

  &-with-children {
    max-width: fit-content;
  }
}

@keyframes bui-skeleton-wave {
  0% {
    background-position: 100% 50%;
  }

  100% {
    background-position: 0 50%;
  }
}

@keyframes bui-skeleton-pulse {
  0% {
    opacity: 1;
  }

  50% {
    opacity: 0.58;
  }

  100% {
    opacity: 1;
  }
}
