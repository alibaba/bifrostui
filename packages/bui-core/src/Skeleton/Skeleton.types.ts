import { OverrideProps } from '@bifrostui/types';
import React from 'react';

export type VariantEnum = 'text' | 'circular' | 'rectangular' | 'rounded';
export type animationEnum = 'wave' | 'pulse' | false;

export type SkeletonProps<
  D extends React.ElementType = 'div',
  P = {},
> = OverrideProps<
  {
    props: P & {
      /**
       * 动效 'wave' | 'pulse' | false
       * 对应动效 波纹 | 脉搏 | 无动效
       * @default 'wave'
       */
      animation?: animationEnum;
      /**
       * 展现形式
       * - text 文本，可以根据`font-size`自动调整高度
       * - circular 圆形
       * - rectangular 方块
       * - rounded 圆角方块
       * @default 'text'
       */
      variant?: VariantEnum;
      /**
       * 占位宽度
       * @default 100%
       */
      width?: string | number;
      /**
       * 占位高度
       */
      height?: string | number;
    };
    defaultComponent: D;
  },
  D
>;
