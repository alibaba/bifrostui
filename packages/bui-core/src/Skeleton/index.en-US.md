---
group: Feedback
name: Skeleton Placeholder
---

# Skeleton Placeholder

Provides a placeholder graphic combination where content needs to be loaded.

## Code Demos

### Basic Usage

Use `width` and `height` to set dimensions.

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack>
      <Skeleton width={210} />
      <Skeleton width={240} />
    </Stack>
  );
};
```

### Variants

The component supports 4 variants `variant`:

- `text` (default): Automatically adjusts size based on `font-size`.
- `circular`, `rectangular`, `rounded`: Adjust size based on `width` and `height`.

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack spacing="10px">
      <Skeleton variant="text" width={210} />
      <Skeleton variant="circular" width={48} height={48} />
      <Skeleton variant="rectangular" width={210} height={40} />
      <Skeleton variant="rounded" width={210} height={40} />
    </Stack>
  );
};
```

### Animation

Default is `wave`. You can set different animations using `animation` or disable it entirely.

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack>
      <Skeleton width={210} />
      <Skeleton animation="pulse" width={210} />
      <Skeleton animation={false} width={210} />
    </Stack>
  );
};
```

### Combinations

You can use different graphic combinations.

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack>
      <div
        style={{ display: 'flex', alignItems: 'center', marginBottom: '12px' }}
      >
        <Skeleton
          variant="circular"
          style={{ margin: '0 6px 0 0', width: '48px', height: '48px' }}
        />
        <div>
          <Skeleton width="50%" style={{ width: '210px' }} />
          <Skeleton width="50%" style={{ width: '210px' }} />
        </div>
      </div>
    </Stack>
  );
};
```

### Size Inference

Dimensions can be set using the `height` and `width` properties of `style`.

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack spacing="10px">
      <Skeleton variant="circular">
        <div style={{ height: '80px', width: '80px' }} />
      </Skeleton>
      <Skeleton variant="circular">
        <div style={{ height: '100px', width: '100px' }} />
      </Skeleton>
    </Stack>
  );
};
```

## API

| Property  | Description        | Type                                               | Default |
| --------- | ------------------ | -------------------------------------------------- | ------- |
| animation | Animation effect   | `wave` \| `pulse` \| `false`                       | `wave`  |
| variant   | Display form       | `text` \| `circular` \| `rectangular` \| `rounded` | `text`  |
| width     | Placeholder width  | `string` \| `number`                               | `100%`  |
| height    | Placeholder height | `string` \| `number `                              | `16px`  |

## Style Variables

| Property | Description        | Default | Global Variable       |
| -------- | ------------------ | ------- | --------------------- |
| --height | Placeholder height | 1.2em   | --bui-skeleton-height |
