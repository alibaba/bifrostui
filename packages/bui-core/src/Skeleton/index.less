@import '@bifrostui/styles/mixins/index.less';

@{root-selector} {
  --bui-skeleton-height: 1.2em;
}

.bui-skeleton {
  display: block;

  &-animation-wave {
    background: linear-gradient(
      90deg,
      rgba(190, 190, 190, 0.2) 25%,
      rgba(129, 129, 129, 0.24) 37%,
      rgba(190, 190, 190, 0.2) 63%
    );
    background-size: 400% 100%;
    animation: bui-skeleton-wave 1.4s ease infinite;
  }

  &-animation-pulse {
    background: rgba(190, 190, 190, 0.35);
    animation: bui-skeleton-pulse 1.4s ease infinite;
  }

  &-text {
    height: auto;
    transform-origin: 0 55%;
    transform: scale(1, 0.6);
    border-radius: var(--bui-shape-radius-label);

    &::before {
      content: '\00a0';
    }
  }

  &-rectangular {
    height: var(--bui-skeleton-height);
  }

  &-rounded {
    height: var(--bui-skeleton-height);
    border-radius: var(--bui-shape-radius-label);
  }

  &-circular {
    border-radius: 50%;
  }

  &-with-children {
    max-width: fit-content;
  }
}

@keyframes bui-skeleton-wave {
  0% {
    background-position: 100% 50%;
  }

  100% {
    background-position: 0 50%;
  }
}

@keyframes bui-skeleton-pulse {
  0% {
    opacity: 1;
  }

  50% {
    opacity: 0.58;
  }

  100% {
    opacity: 1;
  }
}
