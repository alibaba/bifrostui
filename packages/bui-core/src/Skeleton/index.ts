export { default } from './Skeleton';
export { default as Skeleton } from './Skeleton';
export * from './Skeleton.types';
