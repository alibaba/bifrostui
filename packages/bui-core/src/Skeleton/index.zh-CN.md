---
group: 反馈
name: Skeleton 骨架屏
---

# Skeleton 骨架屏

在需要等待加载内容的位置提供一个占位图形组合。

## 代码演示

### 基础用法

使用`width`和`height`设置宽高。

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack>
      <Skeleton width={210} />
      <Skeleton width={240} />
    </Stack>
  );
};
```

### 变体

组件支持 4 种变体`variant`:

`text`(默认): 可以根据`font-size`自动调整大小。

`circular`, `rectangular`, `rounded`: 可以根据`width`和`height`来调整大小。

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack spacing="10px">
      <Skeleton variant="text" width={210} />
      <Skeleton variant="circular" width={48} height={48} />
      <Skeleton variant="rectangular" width={210} height={40} />
      <Skeleton variant="rounded" width={210} height={40} />
    </Stack>
  );
};
```

### 动画

默认是`wave`。你可以根据`animation`来设置不同的动画或者完全禁用它。

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack>
      <Skeleton width={210} />
      <Skeleton animation="pulse" width={210} />
      <Skeleton animation={false} width={210} />
    </Stack>
  );
};
```

### 组合

可以使用不同的图形组合。

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack>
      <div
        style={{ display: 'flex', alignItems: 'center', marginBottom: '12px' }}
      >
        <Skeleton
          variant="circular"
          style={{ margin: '0 6px 0 0', width: '48px', height: '48px' }}
        />
        <div>
          <Skeleton width="50%" style={{ width: '210px' }} />
          <Skeleton width="50%" style={{ width: '210px' }} />
        </div>
      </div>
    </Stack>
  );
};
```

### 尺寸推断

可以根据`style`的`height`和`width`属性来设置尺寸。

```tsx
import React from 'react';
import { Skeleton, Stack } from '@bifrostui/react';

export default () => {
  return (
    <Stack spacing="10px">
      <Skeleton variant="circular">
        <div style={{ height: '80px', width: '80px' }} />
      </Skeleton>
      <Skeleton variant="circular">
        <div style={{ height: '100px', width: '100px' }} />
      </Skeleton>
    </Stack>
  );
};
```

## API

| 属性      | 说明     | 类型                                               | 默认值 |
| --------- | -------- | -------------------------------------------------- | ------ |
| animation | 动画效果 | `wave` \| `pulse` \| `false`                       | `wave` |
| variant   | 展现形式 | `text` \| `circular` \| `rectangular` \| `rounded` | `text` |
| width     | 占位宽度 | `string` \| `number`                               | `100%` |
| height    | 占位高度 | `string` \| `number `                              | `16px` |

## 样式变量

| 属性            | 说明           | 默认值                     |
| --------------- | -------------- | -------------------------- |
| --border-radius | 骨架屏圆角大小 | `--bui-shape-radius-label` |
