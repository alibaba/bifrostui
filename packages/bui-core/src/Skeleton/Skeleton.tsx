import clsx from 'clsx';
import React from 'react';
import './Skeleton.less';
import { SkeletonProps } from './Skeleton.types';

const prefixCls = 'bui-skeleton';

const Skeleton = React.forwardRef<HTMLSpanElement, SkeletonProps>(
  (props, ref) => {
    const {
      className,
      style,
      animation,
      width,
      height,
      variant,
      children,
      ...others
    } = props;

    return (
      <span
        className={clsx(
          prefixCls,
          `${prefixCls}-${variant}`,
          {
            [`${prefixCls}-animation-${animation}`]: animation,
            [`${prefixCls}-with-children`]: Boolean(children),
          },
          className,
        )}
        ref={ref}
        style={{ width, height, ...style }}
        {...others}
      >
        {children}
      </span>
    );
  },
);

Skeleton.displayName = 'BuiSkeleton';
Skeleton.defaultProps = {
  animation: 'wave',
  variant: 'text',
};
export default Skeleton;
