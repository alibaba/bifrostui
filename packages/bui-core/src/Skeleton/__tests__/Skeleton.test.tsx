import React from 'react';
import { render, isConformant, screen } from 'testing';
import Skeleton from '../index';

describe('Skeleton', () => {
  const rootClass = 'bui-skeleton';

  isConformant({
    Component: Skeleton,
    displayName: 'BuiSkeleton',
    className: rootClass,
  });
  it('should render with children', () => {
    render(<Skeleton>Skeleton</Skeleton>);
    expect(screen.getByText('Skeleton')).toHaveClass(
      `${rootClass}-animation-wave`,
    );
    expect(screen.getByText('Skeleton')).toHaveClass(
      `${rootClass}-with-children`,
    );
  });

  it.each(['wave', 'pulse'] as const)(
    'should render correctly with animation props',
    (animation) => {
      render(<Skeleton animation={animation}>Skeleton</Skeleton>);
      expect(screen.getByText('Skeleton')).toHaveClass(
        `${rootClass}-animation-${animation}`,
      );
    },
  );

  it('when animation is false not has animation class', () => {
    render(<Skeleton animation={false}>Skeleton</Skeleton>);
    expect(screen.getByText('Skeleton').getAttribute('class')).not.toMatch(
      /bui-skeleton-animation-/gi,
    );
  });
  it.each(['text', 'circular', 'rectangular', 'rounded'] as const)(
    'should render correctly with variant props',
    (variant) => {
      render(<Skeleton variant={variant}>Skeleton</Skeleton>);
      expect(screen.getByText('Skeleton')).toHaveClass(
        `${rootClass}-${variant}`,
      );
    },
  );

  it('can set width and height in style', () => {
    render(
      <Skeleton width="48px" height="48px">
        Skeleton
      </Skeleton>,
    );
    expect(screen.getByText('Skeleton')).toHaveStyle({
      width: '48px',
      height: '48px',
    });
    expect(screen.getByText('Skeleton')).toMatchSnapshot();
  });
});
