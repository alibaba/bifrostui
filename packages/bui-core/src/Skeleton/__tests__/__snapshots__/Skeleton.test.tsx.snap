// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton can set width and height in style 1`] = `
<span
  class="bui-skeleton bui-skeleton-text bui-skeleton-animation-wave bui-skeleton-with-children"
  style="width: 48px; height: 48px;"
>
  Skeleton
</span>
`;
